apex/normalization/instance_norm.py

```diff
@@ -1,83 +1,594 @@
+        extent,
+        [batch_dim, channel_dim],
+    )
+    grad_scale = (
+        invstd_bcast
+        if weight is None
+        else fd.ops.mul(
+            invstd_bcast,
+            fd.ops.broadcast_in_dim(weight, extent, [0]),
+        )
+    )
+    if use_input_stats:
+        proj = fd.ops.mul(fd.ops.sub(x, mean), proj_scale)
+        grad_input = fd.ops.mul(
+            fd.ops.sub(
+                fd.ops.sub(grad_output, proj),
+                grad_mean,
+            ),
+            grad_scale,
+        )
+    else:
+        grad_input = fd.ops.mul(grad_output, grad_scale)
+
+    if weight is not None:
+        grad_weight = fd.ops.mul(dot_p, invstd)
+        grad_weight_reduced = fd.ops.sum(grad_weight, [0])
+    else:
+        grad_weight_reduced = None
+    if bias is not None:
+        grad_bias = grad_output_sum
+        grad_bias_reduced = fd.ops.sum(grad_bias, [0])
+    else:
+        grad_bias_reduced = None
+
+    return grad_input, grad_weight_reduced, grad_bias_reduced
+
+
+class NormNVFuserFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx: Any,  # contexts are actually objects of the type we are currently defining
+        x: torch.Tensor,
+        weight: Optional[torch.Tensor],
+        bias: Optional[torch.Tensor],
+        running_mean: Optional[torch.Tensor],
+        running_var: Optional[torch.Tensor],
+        use_input_stats: bool,
+        momentum: float,
+        eps: float,
+        unbiased: bool,
+        stat_axes: List[NamedAxis],
+    ) -> torch.Tensor:
+        channels_last = x.is_contiguous(
+            memory_format=torch.channels_last
+        ) or x.is_contiguous(memory_format=torch.channels_last_3d)
+        xorig = x
         if channels_last:
-            order = [0] + [i for i in range(2, len(input.shape))] + [1]
-            _input = input.permute(order)
-        else:
-            _input = input
-        assert _input.is_contiguous()
-        result = instance_norm_nvfuser_cuda.forward(_input, weight, bias, running_mean, running_var,
-                                                    use_input_stats, momentum, eps, channels_last)
-        if len(result) == 3:
-            out, mean, invstd = result
-        else:
-            running_mean, running_var, out, mean, invstd = result
+            order = [0] + [i for i in range(2, len(x.shape))] + [1]
+            x = x.permute(order)
+        assert x.is_contiguous()
+
+        x_datatype = torch2datatype(x.dtype)
+
+        # execute fusion using Python API. Will be cached automatically
+        fs = Fusion()
+        with FusionDefinition(fs) as fd:
+            tv_x = fd.define_tensor(x.ndim, torch2datatype(x.dtype))
+            inputs = [x]
+            if weight is not None:
+                assert bias is not None
+                tv_weight = fd.define_tensor(weight.ndim, torch2datatype(weight.dtype))
+                tv_bias = fd.define_tensor(bias.ndim, torch2datatype(bias.dtype))
```

Comment:
How will this behave when `weight` and `bias` are `None`?

Reply:
In that case, both `tv_weight` and `tv_bias` will be set to `None`. I am not using dummies in this implementation, so they are Python `None`s. I should probably handle `weight` and `bias` separately since someone could just use one or the other though...